test/hotspot/jtreg/serviceability/sa/TestDebugInfoDecode.java

```diff
@@ -0,0 +1,119 @@
+ * @library /test/lib
+ * @modules jdk.hotspot.agent/sun.jvm.hotspot
+ *          jdk.hotspot.agent/sun.jvm.hotspot.code
+ *          jdk.hotspot.agent/sun.jvm.hotspot.debugger
+ *          jdk.hotspot.agent/sun.jvm.hotspot.runtime
+ * @run main/othervm/timeout=2400 -Xmx1g -Xcomp TestDebugInfoDecode
+ */
+
+public class TestDebugInfoDecode {
+
+    private static LingeredApp theApp = null;
+
+    private static void checkDecode(String pid) throws Exception {
+        HotSpotAgent agent = new HotSpotAgent();
+
+        try {
+            agent.attach(Integer.parseInt(pid));
+
+            CodeCacheVisitor v = new CodeCacheVisitor() {
+                    Throwable throwable;
+                    public void prologue(Address start, Address end) {
+                    }
+                    public void visit(CodeBlob blob) {
+                        if (throwable != null) {
+                            // Only report the first failure.
+                            return;
+                        }
+                        if (blob instanceof NMethod) {
+                            NMethod nm = (NMethod) blob;
+                            try {
+                                nm.decodeAllScopeDescs();
+                            } catch (Throwable t) {
+                                System.err.println("Exception while decoding debug info for " + blob);
+                                throwable = t;
+                                throw t;
+                            }
+                        }
+                    }
+                    public void epilogue() {
+                    }
+                };
+            VM.getVM().getCodeCache().iterate(v);
+        } finally {
+            agent.detach();
+        }
+    }
+
+    private static void createAnotherToAttach(long lingeredAppPid) throws Exception {
+        // Start a new process to attach to the lingered app
+        ProcessBuilder processBuilder = ProcessTools.createLimitedTestJavaProcessBuilder(
+            "--add-modules=jdk.hotspot.agent",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.code=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.debugger=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.runtime=ALL-UNNAMED",
+            "TestDebugInfoDecode",
+            Long.toString(lingeredAppPid)
+        );
+        SATestUtils.addPrivilegesIfNeeded(processBuilder);
+        OutputAnalyzer SAOutput = ProcessTools.executeProcess(processBuilder);
+        SAOutput.shouldHaveExitValue(0);
+        System.out.println(SAOutput.getOutput());
+    }
+
+    public static void main (String... args) throws Exception {
+        SATestUtils.skipIfCannotAttach(); // throws SkippedException if attach not expected to work.
+        if (args == null || args.length == 0) {
+            try {
+                theApp = new LingeredApp();
```

Comment:
Is there a reason why previously you had used LingeredAppWithEnum and now you are using LingeredApp?

Reply:
Both were chosen as a result of copy/paste from existing tests and it honestly doesn't matter.  The Xcomp is what's producing nmethods so the actual test doesn't matter.